File: packages/sdk-python/src/driftstack/resources/billing.py
"""Billing resource — /v1/billing (V-082).

``dict[str, Any]`` typing pending the next regen pass.
"""

from __future__ import annotations

from typing import Any

from driftstack.http import AsyncHttpClient, HttpClient
from driftstack.resources._common import coerce_body


class BillingResource:
    """Synchronous billing resource."""

    def __init__(self, http: HttpClient) -> None:
        self._http = http

    def get_state(self) -> dict[str, Any]:
        """Current subscription mirror + trial-pack state."""
        return self._http.request("GET", "/v1/billing")

    def create_checkout_session(self, body: dict[str, Any]) -> dict[str, Any]:
        """Start a paid-tier subscription Checkout session.

        Body shape: ``{"tier": "...", "billing_period": "monthly"|"annual",
        "success_url"?: ..., "cancel_url"?: ...}``.
        """
        return self._http.request(
            "POST", "/v1/billing/checkout-session", json_body=coerce_body(body)
        )

    def start_trial_pack(self, body: dict[str, Any] | None = None) -> dict[str, Any]:
        """Start the $2.99 trial-pack one-time purchase (per ADR-003)."""
        return self._http.request(
            "POST", "/v1/billing/trial-pack", json_body=coerce_body(body or {})
        )

    def create_portal_session(self) -> dict[str, Any]:
        """Open a Stripe Customer Portal session for the calling account."""
        return self._http.request("POST", "/v1/billing/portal-session")


class AsyncBillingResource:
    """Async billing resource."""

    def __init__(self, http: AsyncHttpClient) -> None:
        self._http = http

    async def get_state(self) -> dict[str, Any]:
        return await self._http.request("GET", "/v1/billing")

    async def create_checkout_session(self, body: dict[str, Any]) -> dict[str, Any]:
        return await self._http.request(
            "POST", "/v1/billing/checkout-session", json_body=coerce_body(body)
        )

    async def start_trial_pack(self, body: dict[str, Any] | None = None) -> dict[str, Any]:
        return await self._http.request(
            "POST", "/v1/billing/trial-pack", json_body=coerce_body(body or {})
        )

    async def create_portal_session(self) -> dict[str, Any]:
        return await self._http.request("POST", "/v1/billing/portal-session")

File: packages/api-types/src/billing.ts
// Billing flow schemas (V-082).
//
// Endpoints exposed under /v1/billing/*:
//   - POST /v1/billing/checkout-session   (start a paid-tier subscription)
//   - POST /v1/billing/portal-session     (open Stripe Customer Portal)
//   - GET  /v1/billing/subscription       (current subscription state)
//
// The one-time $2.99 trial_pack was retired 2026-05-27 in favour of a
// perpetual free tier (see AccountTierSchema 'free'); its checkout flow
// and credit state were removed entirely.
//
// At scaffolding time, the actual Stripe API calls are gated behind
// a `BillingProvider` interface so tests run against an in-memory
// provider that returns deterministic checkout URLs / customer IDs.

import { z } from 'zod';
import { AccountTierSchema, Iso8601Schema, PURCHASABLE_TIERS } from './common.js';

// ───────────────────────────────────────────────────────────────────────────
// Checkout session
// ───────────────────────────────────────────────────────────────────────────

export const BillingPeriodSchema = z.enum(['monthly', 'annual']);
export type BillingPeriod = z.infer<typeof BillingPeriodSchema>;

export const CreateCheckoutSessionRequestSchema = z.object({
  /**
   * Target tier. Must be a self-serve paid tier (not 'free' or 'enterprise').
   *
   * V-924 — an enum of the accepted values, not `AccountTierSchema.refine(...)`.
   * A refine is a runtime predicate JSON Schema cannot express, so the generated
   * OpenAPI document emitted all eight tiers and advertised `free` and
   * `enterprise` as valid on a live billing endpoint that returns 400 for both.
   * Same accepted set, same rejection message, accurate published contract.
   */
  tier: z.enum(PURCHASABLE_TIERS, {
    message: 'tier must be a self-serve paid tier (free and enterprise excluded)',
  }),
  billing_period: BillingPeriodSchema,
  /**
   * Where Stripe redirects on success. The `{CHECKOUT_SESSION_ID}` token
   * is replaced server-side. Defaults to the configured success URL when
   * omitted.
   */
  success_url: z.string().url().optional(),
  cancel_url: z.string().url().optional(),
});
export type CreateCheckoutSessionRequest = z.infer<typeof CreateCheckoutSessionRequestSchema>;

export const CreateCheckoutSessionResponseSchema = z.object({
  checkout_url: z.string().url(),
  /** Stripe checkout session id. Echoed for client-side correlation. */
  checkout_session_id: z.string(),
});
export type CreateCheckoutSessionResponse = z.infer<typeof CreateCheckoutSessionResponseSchema>;

// ───────────────────────────────────────────────────────────────────────────
// Customer portal
// ───────────────────────────────────────────────────────────────────────────

export const CreatePortalSessionResponseSchema = z.object({
  portal_url: z.string().url(),
});
export type CreatePortalSessionResponse = z.infer<typeof CreatePortalSessionResponseSchema>;

// ───────────────────────────────────────────────────────────────────────────
// Subscription read
// ───────────────────────────────────────────────────────────────────────────

export const SubscriptionStatusSchema = z.enum([
  'incomplete',
  'incomplete_expired',
  'trialing',
  'active',
  'past_due',
  'canceled',
  'unpaid',
  'paused',
]);
export type SubscriptionStatus = z.infer<typeof SubscriptionStatusSchema>;

export const SubscriptionSchema = z.object({
  tier: AccountTierSchema,
  status: SubscriptionStatusSchema,
  stripe_subscription_id: z.string(),
  current_period_end: Iso8601Schema.nullable(),
  cancel_at_period_end: z.boolean(),
  canceled_at: Iso8601Schema.nullable(),
  created_at: Iso8601Schema,
  updated_at: Iso8601Schema,
});
export type Subscription = z.infer<typeof SubscriptionSchema>;

export const GetBillingStateResponseSchema = z.object({
  subscription: SubscriptionSchema.nullable(),
});
export type GetBillingStateResponse = z.infer<typeof GetBillingStateResponseSchema>;
